File: software/examples/rust/example_all_data.rs
use std::{error::Error, io, thread};
use tinkerforge::{imu_v2_brick::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XXYYZZ"; // Change XXYYZZ to the UID of your IMU Brick 2.0.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let imu = ImuV2Brick::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Create receiver for all data events.
    let all_data_receiver = imu.get_all_data_receiver();

    // Spawn thread to handle received events. This thread ends when the `imu` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for all_data in all_data_receiver {
            println!("Acceleration [X]: {} m/s²", all_data.acceleration[0] as f32 / 100.0);
            println!("Acceleration [Y]: {} m/s²", all_data.acceleration[1] as f32 / 100.0);
            println!("Acceleration [Z]: {} m/s²", all_data.acceleration[2] as f32 / 100.0);
            println!("Magnetic Field [X]: {} µT", all_data.magnetic_field[0] as f32 / 16.0);
            println!("Magnetic Field [Y]: {} µT", all_data.magnetic_field[1] as f32 / 16.0);
            println!("Magnetic Field [Z]: {} µT", all_data.magnetic_field[2] as f32 / 16.0);
            println!("Angular Velocity [X]: {} °/s", all_data.angular_velocity[0] as f32 / 16.0);
            println!("Angular Velocity [Y]: {} °/s", all_data.angular_velocity[1] as f32 / 16.0);
            println!("Angular Velocity [Z]: {} °/s", all_data.angular_velocity[2] as f32 / 16.0);
            println!("Euler Angle [X]: {} °", all_data.euler_angle[0] as f32 / 16.0);
            println!("Euler Angle [Y]: {} °", all_data.euler_angle[1] as f32 / 16.0);
            println!("Euler Angle [Z]: {} °", all_data.euler_angle[2] as f32 / 16.0);
            println!("Quaternion [W]: {}", all_data.quaternion[0] as f32 / 16383.0);
            println!("Quaternion [X]: {}", all_data.quaternion[1] as f32 / 16383.0);
            println!("Quaternion [Y]: {}", all_data.quaternion[2] as f32 / 16383.0);
            println!("Quaternion [Z]: {}", all_data.quaternion[3] as f32 / 16383.0);
            println!("Linear Acceleration [X]: {} m/s²", all_data.linear_acceleration[0] as f32 / 100.0);
            println!("Linear Acceleration [Y]: {} m/s²", all_data.linear_acceleration[1] as f32 / 100.0);
            println!("Linear Acceleration [Z]: {} m/s²", all_data.linear_acceleration[2] as f32 / 100.0);
            println!("Gravity Vector [X]: {} m/s²", all_data.gravity_vector[0] as f32 / 100.0);
            println!("Gravity Vector [Y]: {} m/s²", all_data.gravity_vector[1] as f32 / 100.0);
            println!("Gravity Vector [Z]: {} m/s²", all_data.gravity_vector[2] as f32 / 100.0);
            println!("Temperature: {} °C", all_data.temperature);
            println!("Calibration Status: {:b}", all_data.calibration_status);
            println!();
        }
    });

    // Set period for all data receiver to 0.1s (100ms).
    imu.set_all_data_period(100);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_all_data.go
package main

import (
	"fmt"
	"tinkerforge/imu_v2_brick"
	"tinkerforge/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XXYYZZ" // Change XXYYZZ to the UID of your IMU Brick 2.0.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	imu, _ := imu_v2_brick.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	imu.RegisterAllDataCallback(func(acceleration [3]int16, magneticField [3]int16, angularVelocity [3]int16, eulerAngle [3]int16, quaternion [4]int16, linearAcceleration [3]int16, gravityVector [3]int16, temperature int8, calibrationStatus uint8) {
		fmt.Printf("Acceleration [X]: %f m/s²\n", float64(acceleration[0])/100.0)
		fmt.Printf("Acceleration [Y]: %f m/s²\n", float64(acceleration[1])/100.0)
		fmt.Printf("Acceleration [Z]: %f m/s²\n", float64(acceleration[2])/100.0)
		fmt.Printf("Magnetic Field [X]: %f µT\n", float64(magneticField[0])/16.0)
		fmt.Printf("Magnetic Field [Y]: %f µT\n", float64(magneticField[1])/16.0)
		fmt.Printf("Magnetic Field [Z]: %f µT\n", float64(magneticField[2])/16.0)
		fmt.Printf("Angular Velocity [X]: %f °/s\n", float64(angularVelocity[0])/16.0)
		fmt.Printf("Angular Velocity [Y]: %f °/s\n", float64(angularVelocity[1])/16.0)
		fmt.Printf("Angular Velocity [Z]: %f °/s\n", float64(angularVelocity[2])/16.0)
		fmt.Printf("Euler Angle [X]: %f °\n", float64(eulerAngle[0])/16.0)
		fmt.Printf("Euler Angle [Y]: %f °\n", float64(eulerAngle[1])/16.0)
		fmt.Printf("Euler Angle [Z]: %f °\n", float64(eulerAngle[2])/16.0)
		fmt.Printf("Quaternion [W]: %f\n", float64(quaternion[0])/16383.0)
		fmt.Printf("Quaternion [X]: %f\n", float64(quaternion[1])/16383.0)
		fmt.Printf("Quaternion [Y]: %f\n", float64(quaternion[2])/16383.0)
		fmt.Printf("Quaternion [Z]: %f\n", float64(quaternion[3])/16383.0)
		fmt.Printf("Linear Acceleration [X]: %f m/s²\n", float64(linearAcceleration[0])/100.0)
		fmt.Printf("Linear Acceleration [Y]: %f m/s²\n", float64(linearAcceleration[1])/100.0)
		fmt.Printf("Linear Acceleration [Z]: %f m/s²\n", float64(linearAcceleration[2])/100.0)
		fmt.Printf("Gravity Vector [X]: %f m/s²\n", float64(gravityVector[0])/100.0)
		fmt.Printf("Gravity Vector [Y]: %f m/s²\n", float64(gravityVector[1])/100.0)
		fmt.Printf("Gravity Vector [Z]: %f m/s²\n", float64(gravityVector[2])/100.0)
		fmt.Printf("Temperature: %d °C\n", temperature)
		fmt.Printf("Calibration Status: %b\n", calibrationStatus)
		fmt.Println()
	})

	// Set period for all data receiver to 0.1s (100ms).
	imu.SetAllDataPeriod(100)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
